text/2020-09-24-QoS.md

```diff
@@ -0,0 +1,172 @@
+
+### TiKV Back Pressure
+
+#### Local Back Pressure at TiKV
+
+TiKV will have an admission controller component. This component will track the QoS status and reject queries before they are accepted.
+
+The downside to following this approach is that TiKV does not understand a multi-node query. One node blocking a query can slow down a larger transaction and end up slowing down the system as a whole. Trying to give TiKV global information won’t scale up well for a large cluster.
+
+#### Query inhibition
+
+Queries should be inhibited based on
+* The total capacity available on the node
+* The QoS policies that apply to the query
+* The estimated resources needed for the queries
+
+#### Resource Estimation
+
+The amount of inhibition required depends on the number of requests and amount of resources being requested. Effectively when resources are highly utilized we build up a queue of pending requests with a limited size where the overflow is rejected.
+
+Policy application is allowed to take into account resources that will be used
+
+* less intensive queries can be prioritized above more intensive queries, particularly for bursting
+* queries can be prioritized that together make for better resource utilization given the multiple dimensions of resource usage.
+
+#### Resource measurement
+
+TiKV must measure the resource usage of the node as a whole.
+However, in our first version we do not take into account the actual usage of different policies. To improve our ability to estimate resource usage we will need to develop the ability to measure the actual resources used of policies being applied. These measurements can eventually be used to apply QoS more intelligently. For example, the effects of bad estimates can be corrected.
+
+### QoS Policy
+
+#### QoS Value and composition
+
+QoS is specified as an integer value on a linear scale. A greater value reflects a greater priority and a value twice as large is twice as high of a priority. Negative values are effectively treated as a fraction between 0 and 1.
+
+QoS values can compose in two different ways (these are also discussed in later sections)
+* Inner Override (replace): a table QoS value overrides a keyspace QoS setting
+8 Inner Prioritization (greater specificity): a custom application request QoS value is a priority relative to other application requests. The application as a whole is still governed by the keyspace QoS value
+
+#### QoS Policy stored in PD
+
+A QoS policy is set by an administrator in PD. It is a combination of a region group and a QoS value. The main region group is a key space. Smaller regions within a key space may be specified such as a table and this QoS setting will take precedence over that of the key space. These groups are dynamic (new regions can be added) and translated to regions by PD which has knowledge of tenant and table groupings.
```

Comment:
I am a little fuzzy on this detail, but I know we can now prevent tables from sharing regions, so it should be possible for PD to know this.

Reply:
Perhaps QoS police should bind to a range instead of a some regions, just like placement rules.

Reply:
> Perhaps QoS police should bind to a range instead of some regions, just like placement rules.

This proposal doesn't closely specify how regions will be grouped. Grouping by key range could be a great solution. The problem with this approach is that a region could span multiple key ranges. We could try to use key range as the underlying primitive but also try to have APIs that talk in terms of key spaces. We can also reject key ranges that already don't fully enclose their existing regions. It is noted in placement rules that the key range of a table can change due to DDL commands. So I am thinking that for the first version of QoS PD can understand key spaces but won't understand tables and may need to accept key ranges.